cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,300 @@
+      }
+    }
+    fileMap.clear();
+  }
+
+  private void closeAndDelete(AvroFile avroFile) {
+    try {
+      try {
+        avroFile.close();
+      } finally {
+        if (avroFile.getLocation().exists()) {
+          avroFile.getLocation().delete();
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Error while closing and deleting file {}", avroFile.getLocation(), e);
+    }
+  }
+
+  /**
+   * flushes the contents of all the open avro files
+   * @throws IOException
+   */
+  public void flush() throws IOException {
+    // perform flush on all the files in the fileMap
+    for (AvroFile file : fileMap.values()) {
+      try {
+        file.flush();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing contents to file {}", file.getLocation());
+        throw Throwables.propagate(e);
+      }
+    }
+  }
+
+  /**
+   * Represents an Avro file.
+   *
+   * Since there is no way to check the state of the underlying file on an exception,
+   * all methods of this class assume that the file state is bad on any exception and close the file.
+   */
+  public class AvroFile implements Closeable {
+    private final Location location;
+    private FSDataOutputStream outputStream;
+    private DataFileWriter<GenericRecord> dataFileWriter;
+    private long createTime;
+    private boolean isOpen = false;
+
+    public AvroFile(Location location) {
+      this.location = location;
+    }
+
+    /**
+     * Opens the underlying file for writing.
+     * If open throws an exception then underlying file may still need to be deleted.
+     *
+     * @throws IOException
+     */
+    void open() throws IOException {
+      try {
+        this.outputStream = new FSDataOutputStream(location.getOutputStream(), null);
+        this.dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<GenericRecord>(schema));
+        this.dataFileWriter.create(schema, this.outputStream);
+        this.dataFileWriter.setSyncInterval(syncIntervalBytes);
+        this.createTime = System.currentTimeMillis();
+        // Sync the file as soon as it is created, otherwise a zero length Avro file can get created on OOM
+        sync();
+      } catch (Exception e) {
+        close();
+        throw new IOException("Exception while creating file " + location, e);
+      }
+      this.isOpen = true;
+    }
+
+    public boolean isOpen() {
+      return isOpen;
+    }
+
+    public Location getLocation() {
+      return location;
+    }
+
+    public void append(ILoggingEvent event) throws IOException {
+      try {
+        dataFileWriter.append(co.cask.cdap.logging.serialize.LoggingEvent.encode(schema, event));
+      } catch (Exception e) {
+        close();
+        throw new IOException("Exception while appending to file " + location, e);
+      }
+    }
+
+    public long getPos() throws IOException {
+      try {
+        return outputStream.getPos();
+      } catch (Exception e) {
+        close();
```

Comment:
Same here. Basically never do implicit closing (except in constructor to release resource before throwing, as there will be no object for the caller to call close on).